src/runtime/halide_hexagon_dma_api.cpp

```diff
@@ -0,0 +1,266 @@
+
+/**
+ * halide_hexagon_dmaapp_wrap_buffer
+ * desc: Create Context
+ * Attach the frame to the context
+ * Wrap the device handle over the context
+ */
+int halide_hexagon_dmaapp_wrap_buffer(void *user_context, halide_buffer_t *buf, unsigned char *inframe, bool read, halide_hexagon_dma_user_fmt_t fmt) {
+    halide_assert(user_context, buf->device == 0);
+    if(buf->device != 0) {
+        return -2;
+    }
+
+    halide_assert(user_context, buf->dimensions > 1);
+    halide_assert(user_context, buf->dimensions < 4);
+
+    int num_of_frames = 1;  // ASSUMPTION: step 1 is one input frame only
+    int last_frame = 1;     // ASSUMPTION: always new resources for each frame
+
+    HexagonDmaContext hexagon_dma_context(user_context, num_of_frames);
+    halide_assert(user_context, hexagon_dma_context.context != NULL);
+    dma_context dma_ctxt = hexagon_dma_context.context;
+
+    halide_hexagon_dmart_attach_context(user_context, dma_ctxt, (uintptr_t)inframe, fmt, read, \
+            buf->dim[0].extent, buf->dim[1].extent, buf->dim[1].stride, last_frame);
+
+    buf->device_interface = halide_hexagon_dma_device_interface();
+    buf->device = reinterpret_cast<uint64_t>(dma_ctxt);
+
+    return 0;
+}
+
+/**
+ * halide_hexagon_dmaapp_release_wrapper
+ * desc: Free DMA Engine
+ * Signal to DMA end of frame remove frame references
+ * Free L2 buffer allocations made for transfer
+ * Remove the device handle from the context
+ */
+int halide_hexagon_dmaapp_release_wrapper(void *user_context, halide_buffer_t *buf) {
+    halide_assert(NULL, buf!=NULL);
+    void* handle = NULL;
+
+    //////////////////////////////////////////////////////////////////////
+    dma_context dma_handle = reinterpret_cast<dma_context>(buf->device);
+    uintptr_t  frame = halide_hexagon_dmart_get_frame(user_context, dma_handle);
+    //////////////////////////////////////////////////////////////////////////
+    handle = halide_hexagon_dmart_get_dma_handle(user_context, dma_handle, frame);
+    if(handle == 0) {
+        error(user_context) << "Function failed to get DMA Write  Handle  \n";
+        return HEX_ERROR;
+    }
+
+    if(handle != 0) {
+        dma_finish_frame(handle);
+        uintptr_t fold_addr = halide_hexagon_dmart_get_fold_addr(user_context, dma_handle, frame);
+        halide_hexagon_dma_wrap_memory_free(user_context, dma_handle, (void*)fold_addr);
+        halide_hexagon_dmart_detach_context(user_context, dma_handle, frame);
+        dma_free_dma_engine(handle);
+    }
+    buf->device_interface->impl->release_module();
+    buf->device = 0;
+    buf->device_interface = NULL;
+
+    return HEX_SUCCESS;
+}
+
+/**
+ * halide_hexagon_dmart_get_memory
+ * desc: Calculate width, height, stride and fold from roi buf
+ * Need also type and padding information. How ???
+ * Allocate Memory
+ */
+void* halide_hexagon_dmaapp_get_memory(void* user_context, halide_buffer_t *roi_buf) {
+    void *vret = halide_hexagon_dma_memory_get(user_context, roi_buf);

```

Comment:
Do you think this function will be called from anywhere else? At this point i see that this is the only call site for halide_hexagon_dma_memory_get. If this is going to be the only call-site please consider inlining halide_hexagon_dma_memory_get here. It'll be easier to see what's going on.

Reply:
Also consider returning early if vret is not null. So consider changing this code to
` halide_assert(user_context, roi_buf != NULL);
   if(roi_buf->host != 0) {
        return (void*) roi_buf->host;
    } else {
     }`